File: 684-RedundantConnection/684-RedundantConnection.cpp
// Last updated: 02/08/2025, 18:48:47
class Solution {
public:
    bool dfs(int node, int parent, vector<vector<int>>& adj, vector<bool>& visit){
        if(visit[node]){
            return true;
        }
        visit[node] = true;
        for(auto& neighbor: adj[node]){
            if(neighbor == parent){
                continue;
            }
            if(dfs(neighbor,node,adj, visit)){
                return true;
            }
        }
        return false;
    }
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        // find the nodes
        int n = edges.size();
        vector<vector<int>> adj(n+1);
        for(auto& edge: edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
            vector<bool> visit(n+1,false);
            if(dfs(edge[0],-1,adj,visit)){
                return {edge[0],edge[1]};
            }
        }
        return {};

    }
};

File: 684-RedundantConnection/684-RedundantConnection.py
# Last updated: 01/10/2025, 00:12:42
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]
        
        def union(x,y):
            rootX = find(x)
            rootY = find(y)
            if rootX == rootY:
                return False
            if rank[rootX] > rank[rootY]:
                parent[rootY] = rootX
            elif rank[rootX] < rank[rootY]:
                parent[rootX] = rootY
            else:
                parent[rootY] = rootX
                rank[rootX] += 1
            return True
        
        n = len(edges)
        parent = [i for i in range(n+1)]
        rank = [0] * (n+1)
        for u, v in edges:
            if not union(u,v):
                return [u,v]
        